ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSshell.cpp

```diff
@@ -90,21 +91,25 @@ string execute_shell_for_output(const string &command) {
         TranslateMessage(&msg);
         DispatchMessage(&msg);
       }
+      DWORD dwRead = 0;
+      DWORD dwAvail = 0;
+      PeekNamedPipe(hStdOutPipeRead, NULL, 0, NULL, &dwAvail, NULL);
+      if (dwAvail) {
+        char buffer[dwAvail];
+        BOOL success = ReadFile(hStdOutPipeRead, buffer, dwAvail, &dwRead, NULL);
+        if (success || dwRead) {
+          buffer[dwAvail] = 0;
+          output.append(buffer, dwAvail);
+        }
+      }
     }
     CloseHandle(hStdOutPipeWrite);
     CloseHandle(hStdInPipeRead);
-    char buffer[4096] = { };
-    DWORD dwRead = 0;
-    ok = ReadFile(hStdOutPipeRead, buffer, 4095, &dwRead, NULL);
-    while (ok == TRUE) {
-      buffer[dwRead] = 0;
-      ok = ReadFile(hStdOutPipeRead, buffer, 4095, &dwRead, NULL);
-    }
```

Comment:
Why was this reading twice?

Reply:
That's kinda what I am doing now. That's why I changed it.

Reply:
Josh told me the synchronous version of the API shouldn't have needed a loop, so I still don't get why it was reading twice.

Reply:
The process is garanteed to hang once it gets a certain number of bytes to the buffer, if you try to do it after the process has already quit. It needs to be done during the time it is being run, otherwise the hanging will occur. I am not the only one who has run into this problem 

> The problem is that windows buffers just a limited amount of data. So you must read the pipe while the process is still running, otherwise the process will be blocked because it cannot write any more data to the pipe.

https://stackoverflow.com/a/13894991/4821390

Reply:
See also - https://forum.yoyogames.com/index.php?threads/execute-shell-for-windows-macos-and-ubuntu.7145/post-452964

Reply:
@RobertBColton Alright, I think I addressed the issue. I'm going to test it real quick before I conclude we've made progress. Just want to make sure I did it right.

Reply:
@RobertBColton alright, I tested and it's working ok

Reply:
I understand now what happened, the reading twice is what was causing the freezing because it was blocking waiting to read more bytes than there were available, it's very obvious. Regardless, your solution of obtaining the available bytes is much easier than implementing this as a loop, excellent work.

Reply:
I see, the hang was the result of attempting to read more than the program would ever output.